tests/integrations/raftstore/test_merge.rs

```diff
@@ -1385,3 +1387,140 @@ fn test_merge_pessimistic_locks_when_gap_is_too_large() {
 
     assert_eq!(cluster.must_get(b"k1").unwrap(), b"new_val");
 }
+
+#[test]
+fn test_merge_pessimistic_locks_propose_fail() {
+    let mut cluster = new_server_cluster(0, 2);
+    configure_for_merge(&mut cluster);
+    cluster.cfg.pessimistic_txn.pipelined = true;
+    cluster.cfg.pessimistic_txn.in_memory = true;
+    let pd_client = Arc::clone(&cluster.pd_client);
+    pd_client.disable_default_operator();
+
+    cluster.run();
+
+    cluster.must_transfer_leader(1, new_peer(1, 1));
+
+    cluster.must_put(b"k1", b"v1");
+    cluster.must_put(b"k3", b"v3");
+
+    let region = cluster.get_region(b"k1");
+    cluster.must_split(&region, b"k2");
+    let left = cluster.get_region(b"k1");
+    let right = cluster.get_region(b"k3");
+
+    // Sending a TransferLeaeder message to make left region fail to propose.
+
+    let snapshot = cluster.must_get_snapshot_of_region(left.id);
+    let txn_ext = snapshot.ext().get_txn_ext().unwrap().clone();
+    let lock = PessimisticLock {
+        primary: b"k1".to_vec().into_boxed_slice(),
+        start_ts: 10.into(),
+        ttl: 3000,
+        for_update_ts: 20.into(),
+        min_commit_ts: 30.into(),
+    };
+    assert!(
+        txn_ext
+            .pessimistic_locks
+            .write()
+            .insert(vec![(Key::from_raw(b"k1"), lock)])
+            .is_ok()
+    );
+    // Filter MsgTimeoutNow, so transfer leader will not actually succeed.
```

Comment:
Transfer leader may abort after some tick and accept proposals again which make the case unstable, how about using a fail point instead.

Reply:
I switch to use a fail point instead. I make a more flexible failpoint, so I can also use it in system tests. PTAL again, thanks very much!